### PR TITLE
Parse lat long

### PR DESCRIPTION
**Pull request by TuckerJD**
Resolution of Github Taxonworks Issue #103: New Task: Stepwise lat long extraction from collecting events
Added files:
- controllers/tasks/collecting_events/parse/stepwise/lat_long_controller.rb
- helpers/tasks/collecting_events/parse/stepwise/lat_long_helper.rb
- views/tasks/collecting_events/parse/stepwise/lat_long/index.html.erb
- assets/javascript/views/tasks/collecting_events/parse/stepwise/lat_long/index.js
- views/tasks/collecting_events/parse/stepwise/lat_long/_matching_table.html.erb
- lib/queries/collecting_event_lat_long_extractor_query.rb
- RSpecs for controller and helper

### CODE REVIEW

File: app/controllers/tasks/collecting_events/parse/stepwise/lat_long_controller.rb [hunk 1]
```diff
@@ -0,0 +1,210 @@
...
+      collecting_event_id: collecting_event_id_param,
+      filters:             filters)
+      .all
+      .with_project_id(sessions_current_project_id)
+      .order(:id)
+      .limit(1)
+      .pluck(:id)[0]
+  end
+
+  def current_collecting_event
```
Comment (mjy) [inline app/controllers/tasks/collecting_events/parse/stepwise/lat_long_controller.rb]: Does not belong here.  Put it in a model or helper.

File: app/helpers/tasks/collecting_events/parse/stepwise/lat_long_helper.rb [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+    list    = Utilities::Geo::REGEXP_COORD
+    box_row = ""
+    list.keys.each { |kee|
+      checked = filters.include?(kee)
+      box_row += content_tag(:td, check_box_tag("filters[]", kee.to_s, checked), align: 'center')
+    }
+    box_row.html_safe
+  end
+
+  #{"DD1"=>{},
```
Comment (mjy) [inline app/helpers/tasks/collecting_events/parse/stepwise/lat_long_helper.rb]: Describe what these comments are or remove them.

File: app/models/collecting_event.rb [hunk 3]
```diff
@@ -542,6 +542,12 @@ def data_attributes
       column_names.reject { |c| %w{id project_id created_by_id updated_by_id created_at updated_at project_id}.include?(c) || c =~ /^cached/ }
     end
 
+    # def next_needs_parse(start = nil)
```
Comment (mjy) [inline app/models/collecting_event.rb]: Remove.

File: app/views/tasks/collecting_events/parse/stepwise/lat_long/update.js.erb [hunk 4]
```diff
@@ -0,0 +1,38 @@
...
+function stopSpinner() {
+  timer.stop();
+  $("#matching_span").mx_spinner('hide');
+}
+
+html = $('<%= 'placekeeper' %>');
+startSpinner();
+
+var paging;
+//paging = $('<%#= j(render(partial: 'paging_data')) %>');
```
Comment (mjy) [inline app/views/tasks/collecting_events/parse/stepwise/lat_long/update.js.erb]: Is this a feature waiting to happen, or abandoned?  Likely remove these comments.

File: lib/queries/collecting_event_lat_long_extractor_query.rb [hunk 5]
```diff
@@ -0,0 +1,127 @@
...
+        filter_keys = filters
+      end
+
+      all_filters = filter_keys.collect do |kee|
+        # attach 'verbatim_label ~ ' to each regex
+        regex_function(kee)
+      end.join(' OR ')
+      # remove the names from the named groups
+      Arel.sql("(#{all_filters.gsub('?<lat>', '').gsub('?<long>', '')})")
+      # start with the first scope
```
Comment (mjy) [inline lib/queries/collecting_event_lat_long_extractor_query.rb]: Remove comment block.

File: lib/queries/collecting_event_lat_long_extractor_query.rb [hunk 6]
```diff
@@ -0,0 +1,127 @@
...
+    def table
+      CollectingEvent.arel_table
+    end
+
+    # @return [Scope]
+    def all
+      CollectingEvent.where(where_sql)
+    end
+
+    # @param [String] filter regex pattern for matching lat_lomg
```
Comment (mjy) [inline lib/queries/collecting_event_lat_long_extractor_query.rb]: Remove.

File: lib/utilities/geo.rb [hunk 7]
```diff
@@ -91,6 +91,135 @@ def self.distance_in_meters(elev_in)
...
+          lat_long[:long]  = m[:long]
+        end
+      end
+      lat_long
+    end
+
+    def self.hunt_wrapper(label, filters = REGEXP_COORD.keys)
+
+      trials = self.hunt_lat_long_full(label, filters)
+      # trials                = {}
```
Comment (mjy) [inline lib/utilities/geo.rb]: Remove comments.

File: lib/utilities/geo.rb [hunk 8]
```diff
@@ -91,6 +91,135 @@ def self.distance_in_meters(elev_in)
...
+        trials["(#{sep})"] = trial.merge!(method: "(#{sep})")
+      }
+      trials
+    end
+
+    def self.is_lat_long_special(c)
+      SPECIAL_LATLONG_SYMBOLS.include?(c)
+    end
+
+    def self.guess_lat_long(source_string = '')
```
Comment (mjy) [inline lib/utilities/geo.rb]: Does nothing?  Remove.

File: spec/lib/utilities/geo_spec.rb [hunk 9]
```diff
@@ -138,23 +138,493 @@
...
+          this_case = Utilities::Geo.hunt_wrapper(label)
+          expect(this_case).to eq(result)
+        end
+      }
+    end
+
+    context 'multiple use cases for hunt_wrapper' do
+      use_cases = {
+        "Here is some extra text: N 23.23  W 44.44  More stuff at the end"                                                                                =>
+          {"DD1A" => {:method => "text, DD1A"},
```
Comment (mjy) [inline spec/lib/utilities/geo_spec.rb]: Still don't like capitalized keys.

File: app/controllers/tasks/collecting_events/parse/stepwise/lat_long_controller.rb [hunk 10]
```diff
@@ -0,0 +1,210 @@
...
+    #   if kee.start_with?('select_')
+    #     filters.push(kee.gsub('select_', '').downcase.to_sym)
+    #   end
+    # }
+    # filters
+  end
+
+  # all buttons come here, so we first have to look at the button value
+  def process_buttons
+    prevention = ['39605', '103244', '57187', '103255']
```
Comment (TuckerJD) [inline app/controllers/tasks/collecting_events/parse/stepwise/lat_long_controller.rb]: It's a debugging aid.

File: app/controllers/tasks/collecting_events/parse/stepwise/lat_long_controller.rb [hunk 11]
```diff
@@ -0,0 +1,210 @@
...
+    else
+      flash['alert'] = message
+      next_id        = current_collecting_event.id
+    end
+    # where do we go from here?
+    redirect_to collecting_event_lat_long_task_path(collecting_event_id: next_id,
+                                                    filters:             parse_filters(params))
+  end
+
+  def convert
```
Comment (TuckerJD) [inline app/controllers/tasks/collecting_events/parse/stepwise/lat_long_controller.rb]: It's a landing point (route). Where *does* it belong?

**Comment by mjy** [comment]
My bad, didn't see the return json, it stays.

File: app/controllers/tasks/collecting_events/parse/stepwise/lat_long_controller.rb [hunk 12]
```diff
@@ -0,0 +1,210 @@
...
+                 .with_project_id(sessions_current_project_id)
+                 .order(:id)
+                 .limit(1)
+                 .pluck(:id)[0]
+    end
+    retval
+  end
+
+  # TODO: deprecate for valud from view/helper
+  def next_collecting_event_id
```
Comment (TuckerJD) [inline app/controllers/tasks/collecting_events/parse/stepwise/lat_long_controller.rb]: That's odd; you were the one who put it in the controller....
Also, I have no idea what that TODO was supposed to mean.

File: app/controllers/tasks/collecting_events/parse/stepwise/lat_long_controller.rb (hunk 1 above)
Comment (TuckerJD) [inline app/controllers/tasks/collecting_events/parse/stepwise/lat_long_controller.rb]: That's odd; you were the one who put it in the controller....

**Comment by mjy** [comment]
Is it a route as well?

M

File: spec/lib/utilities/geo_spec.rb (hunk 9 above)
Comment (TuckerJD) [inline spec/lib/utilities/geo_spec.rb]: These is for display purposes, and are used directly.

**Comment by TuckerJD** [comment]
No

_______________________________
James Tucker, Programmer
Species File Group,
Illinois Natural History Survey,
Prairie Research Institute at the University of Illinois, Urbana-Champaign.
TuckerJD@UIUC.Edu<mailto:TuckerJD@UIUC.Edu>
217-244-9416

**Comment by TuckerJD** [comment]
All issues explained or fixed.

**Pull request merged**